headjack/models/utterance.py

```diff
@@ -49,6 +50,26 @@ def convo(
         history = history[::-1]
         return "\n".join(str(u) for u in history) + "\n"
 
+    def _log_str(self) -> str:
+        """Creates a string for logging the utterance"""
+        return json.dumps(
+            {
+                "class": self.__class__.__name__,
+                "id": self.id,
+                "parent": self.parent and self.parent.id,
+                "utterance": self.utterance,
+            },
+        )
+
+    def log(self):
+        """Logs all utterances that have not yet been logged"""
+        if not self._logged():
+            _logger.utterance(self._log_str())  # type: ignore
+            self._set_logged(True)
+        for utterance in self.history():
```

Comment:
Can you explain why the history has to always be checked? A response from an lmql query returns a single utterance right? Can it just be logged at that point?

Reply:
@samredai an agent can add arbitrarily many intermediate values to a chain of utterances